README.rst
Phone-ISO3166
==============================================
Small project to map an E.164 (international) phone number to the
ISO-3166-1 alpha 2 (two letter) country code, associated with that number.

Also provides mapping for E.212 (mobile network codes, mcc+mnc) to the country.

The package has no dependencies, and works on Python 3.6+ and PyPy 3.6+.


Simple Usage
------------

To simply get a country code from a phone number or mcc, mnc.

.. code:: pycon


    >>> from phone_iso3166.country import *
    >>> phone_country(45)
    'DK'
    >>> phone_country('+1 202-456-1111')
    'US'
    >>> phone_country(14412921234)
    'BM'
    >>> network_country(238, 1)
    'DK'


Network names and codes
-----------------------

If you want more detailed information about the networks, such as the operator
name use the functions in `phone_iso3166.network`.

.. code:: pycon


    >>> from phone_iso3166.network import *
    >>> network(238, 1)
    ('DK', 'TDC Mobil')
    >>> country_networks('DK')
    [(238, 1, 'TDC Mobil'), (238, 2, 'Telenor'), (238, 3, 'MIGway A/S'),
     (238, 4, 'NextGen Mobile Ltd T/A CardBoardFish'), (238, 6, 'Hi3G'),
     (238, 8, 'Nordisk Mobiltelefon Danmark A/S'), (238, 10, 'TDC Mobil'),
     (238, 43, 'MobiWeb Limited'), (238, 12, 'Lycamobile Denmark'),
     (238, 13, 'Compatel Limited'), (238, 77, 'Tele2'), (238, 20, 'Telia'),
     (238, 66, 'TT-Netvaerket P/S'), (238, 28, 'CoolTEL'),
     (238, 30, 'Interactive Digital Media GmbH')]


More information
----------------

If want more information, you can easily use the country code with other python
packages such as `pycountry`_.

.. code:: pycon


    >>> from phone_iso3166.country import phone_country
    >>> import pycountry
    >>> phone = '+55 21 3814-2121'
    >>> c = pycountry.countries.get(alpha_2=phone_country(phone))
    >>> c.name
    'Brazil'
    >>> c.official_name
    'Federative Republic of Brazil'


This package makes no attempt to understand the various input options for
phone numbers, and assumes an international phone number. If you deal in fuzzy
inputs, try `phonenumbers`_.

.. code:: pycon


    >>> from phone_iso3166.country import phone_country
    >>> import phonenumbers
    >>> import pycountry
    >>> local = phonenumbers.parse("020 8366 1177", "GB")
    >>> phonenumbers.format_number(local, phonenumbers.PhoneNumberFormat.E164)
    '+442083661177'
    >>> cc = phone_country(str(local.country_code)+str(local.national_number))
    >>> uk = pycountry.countries.get(alpha_2=cc)
    >>> uk.name
    'United Kingdom'
    >>> uk.official_name
    'United Kingdom of Great Britain and Northern Ireland'
    >>> wh = phonenumbers.parse("0012024561111", "GB")
    >>> cc_wh = phone_country(str(wh.country_code)+str(wh.national_number))
    >>> cc_wh
    'US'


.. |build-status| image:: https://travis-ci.org/onlinecity/phone-iso3166.svg?branch=master
   :target: https://travis-ci.org/onlinecity/phone-iso3166
.. |coverage-status| image:: https://img.shields.io/coveralls/onlinecity/phone-iso3166.svg
   :target: https://coveralls.io/r/onlinecity/phone-iso3166
.. _pycountry: https://pypi.python.org/pypi/pycountry
.. _phonenumbers: https://pypi.python.org/pypi/phonenumberslite


CHANGES.rst
Changes
~~~~~~~

0.4.0
-----
- added type hints
- Dropped support for python version below 3.6

0.3.9
-----
- update dataset based on Wikipedia dataset


phone_iso3166/e212.py
# Generated by get_e212_wiki.py
# at 2022-06-22T09:46:51.163246+00:00
#
# Based on the following pages from wikipedia:
# https://en.wikipedia.org/wiki/Mobile_network_codes_in_ITU_region_2xx_(Europe)
# https://en.wikipedia.org/wiki/Mobile_Network_Codes_in_ITU_region_3xx_(North_America)
# https://en.wikipedia.org/wiki/Mobile_network_codes_in_ITU_region_4xx_(Asia)
# https://en.wikipedia.org/wiki/Mobile_Network_Codes_in_ITU_region_5xx_(Oceania)
# https://en.wikipedia.org/wiki/Mobile_Network_Codes_in_ITU_region_6xx_(Africa)
# https://en.wikipedia.org/wiki/Mobile_Network_Codes_in_ITU_region_7xx_(South_America)
#
# Do not make any manual edits to this file.
# Edits will be overwritten when the generator script runs.
import typing

networks: typing.Dict[int, typing.Union[str, typing.Dict[int, str]]] = {
    202: "GR",
    204: "NL",
    206: "BE",
    208: "FR",
    212: "MC",
    213: "AD",
    214: "ES",
    216: "HU",
    218: "BA",
    219: "HR",
    220: "RS",
    221: "XK",
    222: "IT",
    226: "RO",
    228: "CH",
    230: "CZ",
    231: "SK",
    232: "AT",
    234: "GB",
    235: "GB",
    238: "DK",
    240: "SE",
    242: "NO",
    244: "FI",
    246: "LT",
    247: "LV",
    248: "EE",
    250: "RU",
    255: {
        0: "MD",
        1: "UA",
        2: "UA",
        3: "UA",
        4: "UA",
        6: "UA",
        7: "UA",
        10: "UA",
        21: "UA",
        99: "UA",
    },
    257: "BY",
    259: "MD",
    260: "PL",
    262: "DE",
    266: "GI",
    268: "PT",
    270: "LU",
    272: "IE",
    274: "IS",
    276: "AL",
    278: "MT",
    280: "CY",
    282: "GE",
    283: "AM",
    284: "BG",
    286: "TR",
    288: "FO",
    290: "GL",
    292: "SM",
    293: "SI",
    294: "MK",
    295: "LI",
    297: "ME",
    302: "CA",
    308: "PM",
    310: {
        4: "US",
        5: "US",
        6: "US",
        12: "US",
        13: "US",
        15: "US",
        20: "US",
        30: "US",
        32: "US",
        33: "US",
        34: "US",
        35: "US",
        50: "US",
        53: "US",
        54: "US",
        59: "BM",
        66: "US",
        70: "US",
        80: "US",
        90: "US",
        100: "US",
        110: "US",
        120: "VI",
        130: "US",
        140: "US",
        150: "US",
        160: "US",
        170: "US",
        180: "US",
        190: "US",
        260: "VI",
        320: "US",
        330: "US",
        340: "US",
        360: "US",
        370: "US",
        390: "US",
        400: "US",
        410: "VI",
        430: "US",
        440: "US",
        450: "US",
        460: "US",
        470: "US",
        480: "US",
        490: "US",
        500: "US",
        510: "US",
        520: "US",
        530: "US",
        540: "US",
        550: "US",
        570: "US",
        580: "US",
        590: "US",
        591: "US",
        592: "US",
        593: "US",
        594: "US",
        595: "US",
        596: "US",
        597: "US",
        598: "US",
        599: "US",
        600: "US",
        610: "US",
        620: "US",
        640: "US",
        650: "US",
        670: "US",
        680: "US",
        690: "US",
        700: "US",
        710: "US",
        720: "US",
        730: "US",
        740: "US",
        750: "US",
        770: "US",
        810: "US",
        820: "US",
        840: "US",
        850: "US",
        860: "US",
        880: "US",
        890: "US",
        891: "US",
        892: "US",
        893: "US",
        894: "US",
        895: "US",
        896: "US",
        897: "US",
        898: "US",
        899: "US",
        900: "US",
        910: "US",
        920: "US",
        930: "US",
        940: "US",
        950: "US",
        960: "US",
        970: "US",
        990: "US",
    },
    311: {
        0: "US",
        10: "US",
        12: "US",
        20: "US",
        30: "US",
        40: "US",
        50: "US",
        60: "US",
        70: "US",
        80: "US",
        90: "US",
        100: "US",
        110: "US",
        120: "US",
        130: "US",
        140: "US",
        150: "US",
        170: "US",
        190: "US",
        200: "US",
        210: "US",
        220: "US",
        230: "US",
        240: "US",
        250: "US",
        270: "US",
        271: "US",
        272: "US",
        273: "US",
        274: "US",
        275: "US",
        276: "US",
        277: "US",
        278: "US",
        279: "US",
        280: "US",
        281: "US",
        282: "US",
        283: "US",
        284: "US",
        285: "US",
        286: "US",
        287: "US",
        288: "US",
        289: "US",
        320: "US",
        330: "US",
        340: "US",
        350: "US",
        370: "US",
        380: "US",
        390: "US",
        400: "US",
        410: "US",
        420: "US",
        430: "US",
        440: "US",
        450: "US",
        470: "VI",
        480: "US",
        490: "US",
        530: "US",
        540: "US",
        550: "US",
        560: "US",
        570: "US",
        580: "US",
        590: "US",
        600: "US",
        630: "US",
        640: "US",
        650: "US",
        660: "US",
        670: "US",
        680: "US",
        690: "US",
        730: "US",
        740: "US",
        770: "US",
        780: "AS",
        790: "US",
        800: "US",
        810: "US",
        820: "US",
        830: "US",
        840: "US",
        850: "US",
        860: "US",
        870: "US",
        880: "US",
        882: "US",
        890: "US",
        900: "US",
        920: "US",
        930: "US",
        950: "US",
        970: "US",
        980: "US",
        990: "US",
    },
    312: "US",
    313: {
        0: "US",
        10: "US",
        20: "US",
        30: "US",
        40: "US",
        60: "US",
        61: "US",
        70: "US",
        80: "US",
        90: "US",
        100: "US",
        110: "US",
        120: "US",
        130: "US",
        140: "US",
        150: "US",
        160: "US",
        170: "US",
        180: "US",
        190: "US",
        200: "US",
        210: "US",
        220: "US",
        230: "US",
        240: "US",
        250: "US",
        260: "US",
        280: "US",
        290: "US",
        300: "US",
        310: "US",
        320: "US",
        330: "US",
        340: "US",
        370: "US",
        380: "US",
        390: "US",
        400: "US",
        410: "US",
        420: "US",
        440: "US",
        450: "US",
        460: "US",
        470: "US",
        480: "US",
        490: "US",
        500: "US",
        510: "US",
        520: "US",
        540: "US",
        550: "US",
        560: "US",
        570: "US",
        580: "US",
        590: "US",
        600: "US",
        610: "US",
        620: "US",
        630: "US",
        640: "US",
        650: "US",
        660: "US",
        670: "US",
        680: "US",
        690: "US",
        700: "US",
        710: "US",
        720: "US",
        730: "US",
        740: "US",
        750: "US",
        760: "US",
        770: "US",
        790: "PR",
        840: "US",
        850: "US",
        860: "US",
        870: "US",
        880: "US",
        890: "US",
        900: "US",
        910: "US",
        920: "US",
        930: "US",
        940: "US",
        950: "US",
        960: "US",
        970: "US",
        990: "US",
    },
    314: "US",
    315: "US",
    316: "US",
    330: "PR",
    332: "VI",
    334: "MX",
    338: {5: "LC", 50: "TC", 80: "JM", 110: "JM", 180: "JM"},
    340: "GP",
    342: "BB",
    344: "AG",
    346: "KY",
    348: "VG",
    350: "BM",
    352: "GD",
    354: "MS",
    356: "KN",
    358: "LC",
    360: "VC",
    362: {
        31: "BQ",
        33: "CW",
        51: "SX",
        54: "CW",
        59: "BQ",
        60: "BQ",
        63: "CW",
        68: "CW",
        69: "CW",
        74: "CW",
        76: "BQ",
        78: "BQ",
        91: "CW",
        94: "CW",
    },
    363: "AW",
    364: "BS",
    365: "AI",
    366: "DM",
    368: "CU",
    370: "DO",
    372: "HT",
    374: "TT",
    376: "TC",
    400: "AZ",
    401: "KZ",
    402: "BT",
    404: "IN",
    405: "IN",
    406: "IN",
    410: "PK",
    412: "AF",
    413: "LK",
    414: "MM",
    415: "LB",
    416: "JO",
    417: "SY",
    418: "IQ",
    419: "KW",
    420: "SA",
    421: "YE",
    422: "OM",
    424: "AE",
    425: {
        1: "IL",
        2: "IL",
        3: "IL",
        4: "IL",
        5: "PS",
        6: "PS",
        7: "IL",
        8: "IL",
        9: "IL",
        10: "IL",
        12: "IL",
        16: "IL",
        18: "IL",
        19: "IL",
        20: "IL",
        21: "IL",
        22: "IL",
        23: "IL",
        24: "IL",
        25: "IL",
        26: "IL",
        27: "IL",
        28: "IL",
        29: "IL",
    },
    426: "BH",
    427: "QA",
    428: "MN",
    429: "NP",
    430: "AE",
    431: "AE",
    432: "IR",
    434: "UZ",
    436: "TJ",
    437: "KG",
    438: "TM",
    440: "JP",
    441: "JP",
    450: "KR",
    452: "VN",
    454: "HK",
    455: "MO",
    456: "KH",
    457: "LA",
    460: "CN",
    461: "CN",
    466: "TW",
    467: "KP",
    470: "BD",
    472: "MV",
    502: "MY",
    505: "AU",
    510: "ID",
    514: "TL",
    515: "PH",
    520: "TH",
    525: "SG",
    528: "BN",
    530: "NZ",
    536: "NR",
    537: "PG",
    539: "TO",
    540: "SB",
    541: "VU",
    542: "FJ",
    543: "WF",
    544: "AS",
    545: "KI",
    546: "NC",
    547: "PF",
    548: "CK",
    549: "WS",
    550: "FM",
    551: "MH",
    552: "PW",
    553: "TV",
    554: "TK",
    555: "NU",
    602: "EG",
    603: "DZ",
    604: "MA",
    605: "TN",
    606: "LY",
    607: "GM",
    608: "SN",
    609: "MR",
    610: "ML",
    611: "GN",
    612: "CI",
    613: "BF",
    614: "NE",
    615: "TG",
    616: "BJ",
    617: "MU",
    618: "LR",
    619: "SL",
    620: "GH",
    621: "NG",
    622: "TD",
    623: "CF",
    624: "CM",
    625: "CV",
    626: "ST",
    627: "GQ",
    628: "GA",
    629: "CG",
    630: "CD",
    631: "AO",
    632: "GW",
    633: "SC",
    634: "SD",
    635: "RW",
    636: "ET",
    637: "SO",
    638: "DJ",
    639: "KE",
    640: "TZ",
    641: "UG",
    642: "BI",
    643: "MZ",
    645: "ZM",
    646: "MG",
    647: {0: "RE", 1: "YT", 2: "YT", 3: "RE", 4: "RE", 10: "RE"},
    648: "ZW",
    649: "NA",
    650: "MW",
    651: "LS",
    652: "BW",
    653: "SZ",
    654: "KM",
    655: "ZA",
    657: "ER",
    658: "SH",
    659: "SS",
    702: "BZ",
    704: "GT",
    706: "SV",
    708: "HN",
    710: "NI",
    712: "CR",
    714: "PA",
    716: "PE",
    722: "AR",
    724: "BR",
    730: "CL",
    732: "CO",
    734: "VE",
    736: "BO",
    738: "GY",
    740: "EC",
    742: "GF",
    744: "PY",
    746: "SR",
    748: "UY",
    750: "FK",
}
